open_spiel/games/dark_hex.cc

```diff
@@ -0,0 +1,267 @@
+
+#include "open_spiel/games/dark_hex.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "open_spiel/games/hex.h"
+#include "open_spiel/abseil-cpp/absl/strings/str_cat.h"
+#include "open_spiel/spiel_utils.h"
+
+namespace open_spiel {
+namespace dark_hex {
+namespace {
+  
+using hex::kCellStates;
+using hex::kDefaultBoardSize;
+
+using hex::CellState;
+using hex::kMinValueCellState;
+
+using hex::StateToString;
+using hex::PlayerToState;
+
+// Game Facts
+const GameType kGameType{
+  /*short_name=*/"dark_hex",
+  /*long_name=*/"Dark Hex",
+  GameType::Dynamics::kSequential,
+  GameType::ChanceMode::kDeterministic,
+  GameType::Information::kImperfectInformation,
+  GameType::Utility::kZeroSum,
+  GameType::RewardModel::kTerminal,
+  /*max_num_players=*/2,
+  /*min_num_players=*/2,
+  /*provides_information_state_string=*/true,
+  /*provides_information_state_tensor=*/true,
+  /*provides_observation_string=*/true,
+  /*provides_observation_tensor=*/true,
+  /*parameter_specification=*/
+  {
+    {"obstype", GameParameter(static_cast<std::string>(kDefaultObsType))},
+    {"board_size", GameParameter(hex::kDefaultBoardSize)},
+  }
+};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params){
+  return std::shared_ptr<const Game>(
+    new DarkHexGame(params)
+  );
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+} // namespace
+
+DarkHexState::DarkHexState(std::shared_ptr<const Game> game, int board_size,
+                                  ObservationType obs_type)
+    : State(game), state_(game, board_size), board_size_(board_size), 
+                          obs_type_(obs_type) {
+  black_view_.resize(board_size * board_size, CellState::kEmpty);
+  white_view_.resize(board_size * board_size, CellState::kEmpty);
+
+  kNumCells = board_size_ * board_size_;
+  kBitsPerAction = kNumCells + 1;
+  kLongestSequence = kNumCells * 2 - 1;
+}
+
+void DarkHexState::DoApplyAction(Action move) {
+  Player cur_player = CurrentPlayer(); //current player
+  auto& cur_view = (cur_player == Player(0) ? black_view_ : white_view_);
+
+  // Either occupied or not
+  if (state_.BoardAt(move) == CellState::kEmpty) {
+    state_.ApplyAction(move);
+  } else {}
+
+  SPIEL_CHECK_EQ(cur_view[move], CellState::kEmpty);
+  cur_view[move] = state_.BoardAt(move);
+  action_sequence_.push_back(std::pair<int,Action>(cur_player, move));
+}
+
+std::vector<Action> DarkHexState::LegalActions() const {
+  if (IsTerminal()) return {};
+  std::vector<Action> moves;
+  const Player player = CurrentPlayer();
+  const auto& cur_view = (player == Player{0} ? black_view_ : white_view_);
+
+  for (Action move = 0; move < kNumCells; ++move){
+    if (cur_view[move] == CellState::kEmpty){
+      moves.push_back(move);
+    }
+  }
+
+  return moves;
+}
+
+std::string DarkHexState::ViewToString(Player player) const {
+  const auto& cur_view = (player == Player{0} ? black_view_ : white_view_);
+  std::string str;
+
+  // TODO: ??
```

Comment:
What's this for?

Reply:
the todo? It's a reminder to change to the x, y version of the game - get rows and cols instead of a board size if wanted.

Reply:
sure